RFC-002-mapeo-protocol.md

```diff
@@ -0,0 +1,274 @@
+
+This is the channel that is secured using a cryptographic handshake to derive a
+shared ephemeral key (via [NOISE][]), and a bulk encryption algorithm to
+encrypt all of the outgoing data and decrypt incoming data (via [XSalsa20][]).
+
+[NOISE][] was chosen for its properties of providing end-to-end encryption, as
+well as forward secrecy. [Secret Handshake][] was also reviewed, but the
+requirement of knowing each peer's public key in advance did not fit with
+Mapeo's current discovery model, where peers do not need to be known in advance
+to be found on the local network.
+
+[XSalsa20][] was chosen because of its ongoing use in the [hypercore][]
+project, which Mapeo currently depends on for encryption of the hypercore
+protocol.
+
+An encrypted RPC protocol will run over this secure channel, which can support
+subprotocols, such as Hypercore sync, or media blob exchange. All information
+between peers that requires privacy will be passed through this channel.
+
+Finally, if the secure channel cannot be established due to incompatibility
+reasons (such as an update to another encryption algorithm), it should still
+permit the plaintext channel to be established and used. However, if the secure
+channel fails to establish because of security reasons (eg. detecting a
+man-in-the-middle attack), it should abort the connection entirely.
+
+## 3.3 Remote Procedure Call (RPC) System
+
+The [muxrpc][] protocol wire format is documented on the
+[Secure Scuttlebutt Protocol Guide][ssb-proto-rpc].
+
+# 4. Plaintext Channel RPCs
+
+A peer can use this channel to request a listing of app upgrades (not limited
+to its own platform/architecture), download app binaries over it, get peer
+info, and open a secure channel.
+
+This happens on this unencrypted channel so that if/when we need to change the
+encryption protocol or how upgrades are distributed, the upgrade protocol will
+remain unaffected. The goal is to ensure that a very old version of Mapeo can
+still interoperate at the App Upgrade Protocol layer, even if the secure
+protocol's shape has diverged significantly.
+
+Further, plaintext is sufficiently secure for our needs, since application
+binaries are already signed by default for MacOS, Windows, and Android. For
+Linux, an ad-hoc signing mechanism will need to be deployed. The digital
+signature mechanism prevents a man-in-the-middle from modifying the payload
+undetected. The listing of available upgrades will also include a hash, for
+detecting transmission errors.
+
+The upgrade protocol is expressed as an RPC protocol for requests and responses
+between peers. There are three RPCs exposed by this channel:
+
+## 4.1 `GetAvailableBinaries()`
+
+This RPC returns a list of metadata for all application binaries that the peer
+is offering for download. The response will be a JSON list, with each element
+of the form
+
+```js
+{
+  name: "MapeoDesktop-v4.1.3-armv7.zip",
+  type: "application/zip",
+  version: "4.1.3",
+  platform: "desktop",
```

Comment:
If we remove `platform`, the apps will need to know internally which ones refer to mobile and which ones refer to desktop anyways, if we want features like Mapeo Desktop caching mobile upgrades as well, but not e.g. other desktop platforms.

Reply:
I think most of the decisions about what updates to cache would be based on OS, rather than desktop vs mobile. There is no harm in keeping it, but I don't know if it adds anything. It also is potentially confusing where there is not a clear boundary, e.g. tablet/mobile device running windows, or a chrome laptop running Android apps (which you can do). The installers target an OS rather than desktop vs. mobile.

Reply:
I'll leave it in -- I expect it will be useful for e.g. MD deciding which updates for other devices to cache vs not cache.